Add design doc for oe_log_message() callback function (#3342)
openenclave/openenclave: docs/DesignDocs/oe_log_message()_callback_proposal.md

function.

The callback API is supposed to solve such problems.

# User Experience

By adding this public API, a user can customize the output of the log message:
- The output destination can be stdout, stderr or any specified files that the
destination is adjustable during runtime.
- The message can be adjusted according to the log level.
- The user can get at least as much information about the message as internel
API. For example, log time, enclave side or host side, log level.
- Finally, the user can define a function to customize the log message and can
designate the funtion to process all the log messages.

# Specification

Several necessary items should be added to a public header file.
- Since log level is exposed to end users, the related type should also be exposed:
```
typedef enum _oe_log_level
{
    OE_LOG_LEVEL_NONE = 0,
    OE_LOG_LEVEL_FATAL,
    OE_LOG_LEVEL_ERROR,
    OE_LOG_LEVEL_WARNING,
    OE_LOG_LEVEL_INFO,
    OE_LOG_LEVEL_VERBOSE,
    OE_LOG_LEVEL_MAX
} oe_log_level_t;
```
- The message function declaration:
```
typedef void (*oe_log_callback_t)(
    void* context,
    bool is_enclave,
    const char* time,

>     long int usecs,

Reviewer: Can this be struct tm* or something similar?   Don't force conversion to string if the app log will just want it in binary form

Author: Fixed, now the time is passed in as "struct tm*" and the user can define the time printing format. Thanks!

    const char* message);
```
- The necessary global context. Context may be information to be added to message, or
just a parameter used to process the message.
- The function to set up the callback:
```
oe_result_t oe_log_set_callback(void* context, oe_log_callback_t callback)
```

By including the public header file, the end user can perform customized logging.

# Example

Since this api just provides the log callback feature to host application, here
the sample just shows the necessary lines of code at the host side:
```
#include <openenclave/host.h>
#include <openenclave/trace.h>
#include <stdio.h>

// Include the untrusted log_callback header that is generated
// during the build. This file is generated by calling the
// sdk tool oeedger8r against the log_callback.edl file.
#include "log_callback_u.h"

bool check_simulate_opt(int* argc, const char* argv[])
{
    for (int i = 0; i < *argc; i++)
    {
        if (strcmp(argv[i], "--simulate") == 0)
        {
            fprintf(stdout, "Running in simulation mode\n");
            memmove(&argv[i], &argv[i + 1], (*argc - i) * sizeof(char*));
            (*argc)--;
            return true;
        }
    }
    return false;
}

> void customized_log(

Reviewer: Anand asks: 
1) Thread safety: are calls to this callback serialized?  Or can there be multiple threads calling this in parallel?
    Answer: oe_log_message already serializes messages with a mutex lock.
2) when is it safe to call set_callback?

Dave adds to point #2: can set_callback be called after enclave creation?  Can you call it multiple times and change the registered callback during processing?
    Answer: yes.   oe_log_message makes an OCALL and uses whatever callback (if any) is registered at that time

Simon asks: what is the perf overhead of oe_log_message's mutex?
Anand explains mutex is only on host side

Proposal of Public Callback API for oe_log_message()
====

# Motivation

This document describes the proposed public callback API for oe_log_message().

There are several problems with the current logging mechanism that prevent end
users from making full use of logging:
- Message can't be redirected to different destinations on demand.
- Lacking of a method for end users to customize messages.

> - It doesn't have a mechanism to communicate between host application and

Reviewer 2: ```suggestion
- It doesn't have a mechanism to communicate between the host application and
```

> oe_log_message() that end users can't pass more information to the message

Reviewer 2: ```suggestion
oe_log_message() such that end users can't pass more information to the message
```

function.

The callback API is supposed to solve such problems.

# User Experience

By adding this public API, a user can customize the output of the log message:

> - The output destination can be stdout, stderr or any specified files that the

Reviewer 2: ```suggestion
- The output destination can be stdout, stderr, or any specified files that the
```

destination is adjustable during runtime.
- The message can be adjusted according to the log level.

> - The user can get at least as much information about the message as internel

Reviewer 2: ```suggestion
- The user can get at least as much information about the message as the internal
```

> API. For example, log time, enclave side or host side, log level.

Reviewer 2: ```suggestion
API. For example, log time, enclave or host side, and log level.
```

    if (result != OE_OK)
    {
        fprintf(
            stderr,
            "oe_create_log_callback_enclave(): result=%u (%s)\n",
            result,
            oe_result_str(result));
        goto exit;
    }

    // Call into the enclave
    result = enclave_log_callback(enclave);
    if (result != OE_OK)
    {
        fprintf(
            stderr,
            "calling into enclave_log_callback failed: result=%u (%s)\n",
            result,
            oe_result_str(result));
        goto exit;
    }

    ret = 0;

exit:
    // Clean up the enclave if we created one
    if (out_file)
        fclose(out_file);

    if (enclave)
        oe_terminate_enclave(enclave);

    return ret;
}
```
First the user may prepare a customized log processing function
"customized_log()", which must follow the function type "\*oe_log_callback_t".
All these parameters are some information about logging, but the user has the
full control to make use of them. The timestamp information, which is not
printable, should be converted to a string. Here the printing format
"%Y-%m-%dT%H:%M:%S%z" is a good format that it follows the [ISO 8601 rules](https://en.wikipedia.org/wiki/ISO_8601).

> After that the printing destination is choosed, based on the log level.

Reviewer: ```suggestion
After that, the printing destination is chosen, based on the log level.
```

> Principally we'd better not to mess low priority messages with high priority

Reviewer: ```suggestion
In principle, it's better not to mix low priority messages with high priority
```

> messages in case important messages be flooded by tedious messages.

Reviewer: ```suggestion
messages lest important messages be flooded by tedious messages.
```

> Finally all the informaton and messages will be writen to the specified

Reviewer 2: ```suggestion
Finally, all the information and messages will be written to the specified
```

destination.

> The user can register the customized log function at any time before
> oe_log_message() is called. Obviously if it is registered  before starting

Reviewer: ```suggestion
When oe_log_message() is called, the most recently registered custom log function will be
invoked. Obviously, if it is registered before starting
```

the enclave, all the logs are affected.

To pass in the specified log file, the context is assigned as a pointer to the
desired file "./oe_out.txt". In this case the context is used as the logging
destination. Besides that context can be casted back as any desired type, since
type of context is "void\*".

In the user's expectation, log messages with level equal or greater than

> OE_LOG_LEVEL_WARNING will be writen to "./oe_out.txt", else will be writen to

Reviewer: ```suggestion
OE_LOG_LEVEL_WARNING will be writen to "./oe_out.txt", and others will be writen to
```

Proposal of Public Callback API for oe_log_message()
====

# Motivation

This document describes the proposed public callback API for oe_log_message().

> There are several problems about the current logging that prevent end users to

Reviewer 2: ```suggestion
There are several problems about the current logging mechanism that prevent end users to
```

Author: Updated, thanks!

> There are several problems about the current logging that prevent end users to
> make full use of log:

Reviewer: ```suggestion
There are several problems with the current logging mechanism that prevent end users from
making full use of logging:
```

Author: Updated, thanks!

> - It does't have mechanism to communicate between enclave and host.

Reviewer 3: The current log calls printf, so that is an ocall from enclave to host. Maybe you can elaborate what you mean over here.

Reviewer 2: Echo @radhikaj. Can you elaborate on `mechanism to communicate between enclave and host`?

Author: Updated, thanks!

> - It does't have mechanism to communicate between enclave and host.

Reviewer: ```suggestion
- It doesn't have a mechanism to communicate between enclave and host.
```

Author: Updated, thanks!

- Message can't be redirected to different destinations on demand.

> - Lack of method for end user to customize messages.

Reviewer 2: ```suggestion
Lacking of a method for end users to customize messages.
```

Author: Updated, thanks!

The callback API is supposed to solve such problems.

# User Experience

> 

Reviewer 2: It's not clear to me how the above three problems are addressed based on the current context. It would be good to re-structure a bit.

Author: I have added some sentence and also a sample, thanks!

> By adding this public API, user can customize the output of the log message:

Reviewer 2: ```suggestion
By adding this public API, a user can customize the output of the log message:
```

Author: Updated, thanks!

> - The output destination can be stdout/stderr or any specified files, and it is

Reviewer 2: ```suggestion
- The output destination can be stdout, stderr, or any specified files and the destination
```

Author: Updated, thanks!

> - The output destination can be stdout/stderr or any specified files, and it is

Reviewer: re: "files" plural, does it really support logging the same message to multiple files?

Author: Yes, the user can pass in multiple files if the context is a array of files, and the user have full control to write filtered messages to one specified file, or multiple files as he wish.

adjustable during runtime.
- The message can be adjusted according to the log level.

> - The user can get as much information about the message as internel API. For

Reviewer 2: Does it just mean that there is no information gap between internal and public logging anymore?

Author: Yes all the information are exposed to end users.

example, log time, enclave side/host side, log level.

> - Finally user can define a function to customize the log message, and can

Reviewer 2: ```suggestion
- Finally, the user can define a function to customize the log message and can
```

> designate such funtion to process all the log message.

Reviewer 2: ```suggestion
designate the function to process all the log messages.
```

Author: Updated, thanks!

# Specification

> Several neccessary items should be added to public header file.

Reviewer 2: ```suggestion
Several necessary items should be added to a public header file.
```

Author: Updated, thanks!

> Several neccessary items should be added to public header file.

Reviewer: blocking issue: Which public header file?  Is this a host-side or an enclave-side API or both?

Author: This is a host-side API, the planned header file is include/openenclave/trace.h

- Since log level is exposed to user, the related type should also be exposed:

> `

Reviewer 4: When writing blocks of code, you should use the triple backticks instead, otherwise everything renders as a single line. Doing so will also allow you to specify the type of information it contains for syntax highlighting:  
````md
```c
typedef enum _oe_log_level
{
    OE_LOG_LEVEL_NONE = 0,
    OE_LOG_LEVEL_FATAL,
    OE_LOG_LEVEL_ERROR,
    OE_LOG_LEVEL_WARNING,
    OE_LOG_LEVEL_INFO,
    OE_LOG_LEVEL_VERBOSE,
    OE_LOG_LEVEL_MAX
} oe_log_level_t;
```
````
will render as:
```c
typedef enum _oe_log_level
{
    OE_LOG_LEVEL_NONE = 0,
    OE_LOG_LEVEL_FATAL,
    OE_LOG_LEVEL_ERROR,
    OE_LOG_LEVEL_WARNING,
    OE_LOG_LEVEL_INFO,
    OE_LOG_LEVEL_VERBOSE,
    OE_LOG_LEVEL_MAX
} oe_log_level_t;
```
instead of how it looks right now: `typedef enum _oe_log_level
{
    OE_LOG_LEVEL_NONE = 0,
    OE_LOG_LEVEL_FATAL,
    OE_LOG_LEVEL_ERROR,
    OE_LOG_LEVEL_WARNING,
    OE_LOG_LEVEL_INFO,
    OE_LOG_LEVEL_VERBOSE,
    OE_LOG_LEVEL_MAX
} oe_log_level_t;
`

Author: Updated, Thanks for the suggestion!

typedef enum _oe_log_level
{
    OE_LOG_LEVEL_NONE = 0,
    OE_LOG_LEVEL_FATAL,
    OE_LOG_LEVEL_ERROR,
    OE_LOG_LEVEL_WARNING,
    OE_LOG_LEVEL_INFO,
    OE_LOG_LEVEL_VERBOSE,
    OE_LOG_LEVEL_MAX
} oe_log_level_t;
`
- The message function declaration:
`
typedef void (*oe_log_callback_t)(
    void* context,
    bool is_enclave,
    const char* time,
    long int usecs,

>     oe_log_level_t level,

Reviewer 3: So based on this log level, messages are filtered and the callback occurs only based on the message meeting the log level bar?

Author: So far all the messages will be processed by the callback function if it is available. Users may add filter based on level in the function.

    uint64_t host_thread_id,
    const char* message);
`

> - The necessary global variables.

Reviewer 3: What are the global variables?

Author: Updated, thanks!

- The function to set up the callback:
`
oe_result_t oe_log_set_callback(void* context, oe_log_callback_t callback)
`

> By including the public header file, end user can define the customized message

Reviewer 3: ```suggestion
By including the public header file, the end user can perform customized logging.
```

Author: Updated, now it is not so tedious, thanks!

> and use it to process the log message at runtime.

Reviewer 2: What does `it` refer to?

Author: Updated, thanks!
